nucliadb/nucliadb/ingest/orm/processor.py

```diff
@@ -471,6 +448,47 @@ async def txn(
             counter=counter,
         )
 
+    async def index_resource(
+        self,
+        resource: Resource,
+        txn: Transaction,
+        kb: KnowledgeBox,
+        uuid: str,
+        seqid: int,
+        partition: str,
+    ) -> Tuple[Optional[ShardCounter], Shard]:
+        shard_id = await kb.get_resource_shard_id(uuid)
+        node_klass = get_node_klass()
+        shard: Optional[Shard] = None
+
+        if shard_id is not None:
+            shard = await kb.get_resource_shard(shard_id, node_klass)
+
+        if shard is None:
+            # It's a new resource, get current active shard to place
+            # new resource on
+            shard = await node_klass.get_current_active_shard(txn, kb.kbid)
+            if shard is None:
+                # No shard available, create a new one
+                similarity = await kb.get_similarity()
+                shard = await node_klass.create_shard_by_kbid(
+                    txn, kb.kbid, similarity=similarity
+                )
+            await kb.set_resource_shard_id(uuid, shard.sharduuid)
+
+        if shard is None:
+            raise AttributeError("Shard not available")
+
+        counter = await shard.add_resource(
+            resource.indexer.brain, seqid, partition=partition, kb=kb.kbid
+        )
+        if counter is not None and counter.fields > settings.max_shard_fields:
```

Comment:
While looking at this, I just want to mention that I wonder if we should change how we create new shards.

`add_resource` should not also being calling `GetCount` and returning a counter -- seems very odd

This could be on a separate asyncio queue reader so we're not doing this request during writes